Using ephemeral resource (in provider acceptance test) causes non-refresh plan to be not empty

### Community note

> [!TIP]
> 👋 Hi there, OpenTofu community! The OpenTofu team prioritizes issues based on upvotes. Please make sure to upvote this issue and describe how it affects you in detail in the comments to show your support.


### OpenTofu Version

```shell
OpenTofu v1.11.5
on darwin_arm64
```

### OpenTofu Configuration Files

Minimal reproducible configuration:

```hcl
terraform {
  required_providers {
    random = {
      source = "hashicorp/random"
      version = "3.8.1"
    }
  }
}

provider "random" {}

ephemeral "random_password" "password" {
  length           = 16
}
```

Acceptance test configuration example where ephemeral resource is used to configure `kubernetes` provider: [UpCloudLtd/terraform-provider-upcloud : upcloud/testdata/upcloud_kubernetes/kubernetes_e2e.tf](https://github.com/UpCloudLtd/terraform-provider-upcloud/blob/3ce777953384d0342dd87e28d7b6f3d52b99ca4b/upcloud/testdata/upcloud_kubernetes/kubernetes_e2e.tf)

### Debug Output

Related action in JSON plan file:

```json
{
  "address": "ephemeral.random_password.password",
  "mode": "ephemeral",
  "type": "random_password",
  "name": "password",
  "provider_name": "registry.opentofu.org/hashicorp/random",
  "change": {
    "actions": [
      "open"
    ],
    "before": null,
    "after": null,
    "after_unknown": {},
    "before_sensitive": false,
    "after_sensitive": false
  }
}
```

### Expected Behavior

None or only no-op actions in plan: no need to open the ephemeral resource.

### Actual Behavior

There is a planned action to open the ephemeral resource (even if all other actions are no-op).

### Steps to Reproduce

1. `tofu init`
2. `tofu apply`
3. `tofu plan -refresh=false -out=ephemeral.plan`
4. `tofu show -json -plan=ephemeral.plan`

### Additional Context

When present in provider acceptance test, this planned open action for ephemeral resource causes the test to fail with:

```txt
=== NAME  TestEndToEndKubernetes
    resource_upcloud_kubernetes_test.go:244: Step 1/2 error: After applying this test step, the non-refresh plan was not empty.
        stdout:
        
        
        No changes. Your infrastructure matches the configuration.
        
        OpenTofu has compared your real infrastructure against your configuration and
        found no differences, so no changes are needed.
--- FAIL: TestEndToEndKubernetes (940.19s)
```

### References

_No response_

--- a/website/docs/language/providers/index.mdx
+++ b/website/docs/language/providers/index.mdx
@@ -106,3 +106,55 @@ TF_ACC_TERRAFORM_PATH="/path/to/opentofu"
 TF_ACC_PROVIDER_NAMESPACE="hashicorp"
 TF_ACC_PROVIDER_HOST="registry.opentofu.org"
 ```
+
+#### Known Divergences
+
+##### Ephemeral Resources in Plan Output
+
+OpenTofu includes [ephemeral resources](../../language/ephemerality/ephemeral-resources.mdx) with an `open` action in the JSON plan output's `resource_changes`. This is needed internally for building the apply execution graph when running `tofu apply <planfile>`.
+
+OpenTofu's internal plan emptiness check correctly ignores `open` actions — so the CLI output will say "No changes" and `-detailed-exitcode` returns `0`. However, the [`terraform-plugin-testing`](https://github.com/hashicorp/terraform-plugin-testing) framework interprets any entry in `resource_changes` as a non-empty plan, which can cause acceptance test failures when ephemeral resources are present.
+
+The recommended workaround is to conditionally set `ExpectNonEmptyPlan` when running against OpenTofu. First, add a helper to detect OpenTofu:
+
+```go
+// UsingOpenTofu reports whether the acceptance test binary is OpenTofu.
+// OpenTofu includes ephemeral resource "open" actions in the JSON plan's
+// resource_changes, which causes terraform-plugin-testing to consider the
+// plan non-empty. See: https://github.com/opentofu/opentofu/issues/3781
+func UsingOpenTofu() bool {
+	return strings.HasSuffix(os.Getenv("TF_ACC_TERRAFORM_PATH"), "tofu")
+}
+```
+
+Then on any test step that includes ephemeral resources:
+
+```go
+{
+	Config:             testConfig,
+	ExpectNonEmptyPlan: UsingOpenTofu(),
+},
+```
+
+If you run acceptance tests against both OpenTofu and Terraform, you can use `SkipFunc` to define separate steps for each:
+
+```go
+isTofu := UsingOpenTofu()
+
+// Step for OpenTofu
+{
+	Config:             testConfig,
+	ExpectNonEmptyPlan: true,
+	SkipFunc: func() (bool, error) {
+		return !isTofu, nil
+	},
+},
+// Step for Terraform
+{
+	Config:             testConfig,
+	ExpectNonEmptyPlan: false,
+	SkipFunc: func() (bool, error) {
+		return isTofu, nil
+	},
+},
+```

